sbncode/TPCReco/VertexStub/VertexStubTracker_module.cc

```diff
@@ -181,95 +120,71 @@ void sbn::VertexStubTracker::produce(art::Event& e)
+    if (vhits.size()) vertex = *vhitVtxs.at(0).at(0);
+  
+    std::vector<sbn::StubInfo> stubs;
+    for (unsigned i_vhit = 0; i_vhit < vhits.size(); i_vhit++) {
+      // Collect data on this Vertex-Hit
+      const sbn::VertexHit &thisVHit = *vhits[i_vhit];
+      const recob::Hit &thisVHitHit = *vhitHits.at(i_vhit).at(0);
 
-    art::FindManyP<recob::Hit> findThisSliceHits({thisSlice}, e, fPFPLabel);
-    const std::vector<art::Ptr<recob::Hit>> &hits = findThisSliceHits.at(0);
+      bool passcut = (thisVHit.dqdx >= fdQdxCut) || ((abs(thisVHit.wire.Wire - thisVHit.vtxw) < 1.) && (thisVHit.dqdx >= fOneWiredQdxCut));
 
-    // If we haven't seen this slice beore, collect the hits for its pfparticles
-    if (!slicePFPHits.count(thisSlice.key())) {
-      art::FindManyP<recob::Cluster> pfparticleClusters(thisSlicePFPs, e, fPFPLabel);
+      if (!passcut) continue;
 
-      std::vector<std::vector<art::Ptr<recob::Hit>>> thisSlicePFPHits;
-      for (unsigned i_slc_pfp = 0; i_slc_pfp < thisSlicePFPs.size(); i_slc_pfp++) {
-        thisSlicePFPHits.emplace_back();
-        const std::vector<art::Ptr<recob::Cluster>> &thisPFPClusters = pfparticleClusters.at(i_slc_pfp);
-        art::FindManyP<recob::Hit> clusterHits(thisPFPClusters, e, fPFPLabel);
-        for (unsigned i_clus = 0; i_clus < thisPFPClusters.size(); i_clus++) {
-          thisSlicePFPHits.back().insert(thisSlicePFPHits.back().end(), clusterHits.at(i_clus).begin(), clusterHits.at(i_clus).end());
-        }
-      }
-      slicePFPHits[thisSlice.key()] = thisSlicePFPHits;
-    }
+      sbn::StubInfo sinfo;
+      sinfo.stub = fStubBuilder.FromVertexHit(thisSlice, thisVHit, thisVHitHit, vertex, geo, sce, clock_data, dprop, sinfo.hits, sinfo.pfp); 
+      sinfo.vhit = vhits[i_vhit];
+      sinfo.vhit_hit = vhitHits.at(i_vhit).at(0);
 
-    const std::vector<std::vector<art::Ptr<recob::Hit>>> &thisSlicePFPHits = slicePFPHits.at(thisSlice.key());
+      stubs.push_back(sinfo);
 
-    const recob::Vertex &vert = *vhitVtxs.at(i_vhit).at(0);
-    TVector3 vert_v(vert.position().X(), vert.position().Y(), vert.position().Z());
+    } // end iterate over vertex hits
 
-    std::vector<art::Ptr<recob::Hit>> thisHits = CollectHits(hits, thisVHit, thisVHitHit, vert.position(), geo, dprop);
-
-    // Find which pfparticle to group this VertexStub with 
-    art::Ptr<recob::PFParticle> thisStubPFP = FindPFP(thisHits, thisSlicePFPs, thisSlicePFPHits);
-
-    float charge = 0.;
-    std::set<int> wires;
-    for (unsigned i_hit = 0; i_hit < thisHits.size(); i_hit++) {
-      const recob::Hit &hit = *thisHits[i_hit];
-      charge += fCaloAlg.ElectronsFromADCArea(hit.Integral(), hit.WireID().Plane) * fCaloAlg.LifetimeCorrection(clock_data, dprop, hit.PeakTime(), 0.);
-      wires.insert(hit.WireID().Wire);
+    // Run all of the merging tools
+    for (unsigned i_mrg = 0; i_mrg < fStubMergeTools.size(); i_mrg++) {
+      stubs = fStubMergeTools[i_mrg]->Merge(stubs, geo, sce, clock_data, dprop);
```

Comment:
Usual suggestion to use range-for loop to avoid index mistakes:
```suggestion
    for (auto const& mergeTool: fStubMergeTools) {
      stubs = mergeTool->Merge(stubs, geo, sce, clock_data, dprop);
```

Reply:
I think it's fine as is -- there are no mistakes made with the index here.